src/Sylius/Component/Inventory/Operator/BackordersHandler.php

```diff
@@ -78,16 +86,19 @@ public function fillBackorders(StockableInterface $stockable)
             return;
         }
 
-        $units = $this->repository->findBy([
-            'stockable' => $stockable,
-            'inventoryState' => InventoryUnitInterface::STATE_BACKORDERED,
-        ], ['createdAt' => 'ASC']);
+        $units = $this->repository->findByStockableAndInventoryState(
+            $stockable,
+            InventoryUnitInterface::STATE_BACKORDERED,
+            $onHand
+        );
 
         foreach ($units as $unit) {
-            $unit->setInventoryState(InventoryUnitInterface::STATE_SOLD);
+            $sm = $this->factory->get($unit, InventoryUnitTransitions::GRAPH);
```

Comment:
I think this line can be written simpler, since you do not need the `if` that is below:
```php
$this->factory->get($unit, InventoryUnitTransitions::GRAPH)->apply(InventoryUnitTransitions::SYLIUS_HOLD);
```